### DISCUSSION TITLE
v2.0.0-openbeta01について気づいたことがあれば

### DISCUSSION THREAD

**Discussion by keiji** [cocoa-mhlw/cocoa#921] [General]
v2.0.0-openbeta01（オープンベータ）に関するDiscussionです。
iOS版、Android版ともに配信開始済。

オープンベータには次のURLから参加できます。

* Android版 : https://play.google.com/apps/testing/jp.go.mhlw.covid19radar
* iOS版 : https://testflight.apple.com/join/2tBJvty4

詳細（既知の不具合含む）は次のドキュメントを参照してください。
https://cocoa-mhlw.github.io/cocoa/docs/contribution/open_beta_version/

**Comment by b-wind** [comment]
β版ユーザーがオープンβ版から既存リリース版・将来的なリリース版への移行する際の手順、データの引き継ぎがあるか等は明示した方が良いのかなと。
（技術的には出来てしまうけれど、やらないで欲しい事等も）


個人的には移行をサポートするコスト・リスクは高いので原則非対応。アプリ消去しての再インストールのみ想定という所でしょうか。

**Reply by keiji** [reply]
はい。基本的にアプリ消去から再インストールと言う流れを想定しています。
とはいえ、書いておいた方がいいのは確かですね。あとでIssue化します。ありがとうございます。

ちなみに、アプリを消去してすぐにインストールすればENの情報を引き継げるらしいことを確認していますが、何分までならOKとかはまだ分かっていないですね……

**Comment by okumuralab** [comment]
`ExposureChecks-*.json` から `MatchCount` がなくなっているみたいですが，これでいいんでしたっけ？

```
{
  "Build" : "19E241",
  "ExportVersion" : 2,
  "ExposureChecks" : [
    {
      "Timestamp" : "2022-03-19 03:25:46 +0900",
      "Files" : [
        {
          "Hash" : "B50B03128A719CF24B218A10573FF770D2AE1FFA08CEE424E7A9382E4E33865B",
          "KeyCount" : 3,
          "AppBundleIdentifier" : "jp.go.mhlw.covid19radar",
          "Timestamp" : "2022-03-19 03:25:46 +0900"
        },
```

**Reply by keiji** [reply]
他の方からもTwitterで指摘もらってるのですが、これはENv2に更新したことでiOSのシステムがログの出力内容を変えたのだと思います。
開発チームや行政・連携チームが `MatchCount` を非表示にしているとかではありません（ここはアプリから制御できる場所ではないので）。

COCOAログチェッカーとかよく使われているのは承知していますし、MatchCountがなくなるのは不便ですね……実は閾値未満の接触画面というの企画していて、実装もできています。

https://github.com/cocoa-mhlw/cocoa/issues/823

こちらが搭載されれば解決なのですが、行政内での調整が付かずオープンベータ・リリース版ともにも搭載は未定です（現時点でClosed betaのみに搭載）。少し考えないといけませんね。

**Reply by okumuralab** [reply]
なるほど，ENv2の仕様なんですね。閾値未満の接触が何らかの形で見えるようになれば，COCOAが周囲で使われていることが実感できそうに思いました。

**Reply by keiji** [reply]
閾値未満の接触画面が搭載されれば解決するのですが、搭載は未定というのが現状です。
COCOA2へのバージョンアップの結果として、利用者に提供される情報が減ってしまうと言うのは課題ですね。

あくまでアイデアベースですが、接触情報自体をアプリから取り出せるようにする（そういうメニューを作る）というのも考えられますね。
端末に保存している接触情報そのものは本来、利用者のものなので、利用者からアクセスする方法があることがフェアだと個人的には考えています。

**Reply by okumuralab** [reply]
あと，MatchCountがなくなった後でも，「一致したキーの数」は0が表示され続けるんでしょうか？

![IMG_7586](https://user-images.githubusercontent.com/3616498/159193045-df30a24b-cbe2-4ff2-9ec8-0d27627ca13b.PNG)

**Reply by keiji** [reply]
そこは表示され続ける。一致したキーがあれば、その値が表示されると理解しています。
ENv2ではJSONに出力されなくなったという位置づけでしょうか。

**Reply by okumuralab** [reply]
あぁそうなんですね（確かに出力欄があるのに出力されないというのは変ですよね）。ありがとうございます。

**Reply by daisuke-nogami** [reply]
「一致したキーの数」は表示されるはずです。
数日前、手元の一緒に持ち歩いていた、オープンβ入れてるiPhoneと、オープンβ入れてないAndroid端末とほぼ同じタイミングで「一致したキー」がでたもので…

**Reply by b-wind** [reply]
> 閾値未満の接触が何らかの形で見えるようになれば，COCOAが周囲で使われていることが実感できそうに思いました。

こちらに関しては、できる限りCOCOA自体で表示したいところですね。
OS のログ出力とは見ている物が違うので現状のCOCOAログチェッカーとは少し意味合いが異なるかも知れませんが、需要が大きい事には違い無いので。

**Reply by b-wind** [reply]
> 端末に保存している接触情報そのものは本来、利用者のものなので、利用者からアクセスする方法があることがフェア

具体的にどのような情報を取得しているかの透明性確保の為にも情報のエクスポート機能の様な物は有った方が良いのかなと感じてはいます。
細かい所までCOCOA自体で表示出来た方が良いとは思いますが、色々なパターンの需要にすべて対応するのも難しいと言う側面もありますね。

表示する項目自体は検討の余地があるとは思いますが、あくまで自身の情報を見ると言う点で考えれば問題になることは恐らく無さそうに思います。

**Reply by keiji** [reply]
Issue作りました。本件、こちらで検討していきたいと思います。
https://github.com/cocoa-mhlw/cocoa/issues/928

**Reply by okumuralab** [reply]
たいへん遅ればせながら，ENv2になってもiPhoneの「設定」→「接触通知」で「一致したキーの数」がちゃんと表示されるかという話に関して，私のiPhoneでもちゃんと表示されました。ただし，非減少なので，積分表示かもしれないです。

![matchcount](https://user-images.githubusercontent.com/3616498/161867348-dff68952-9c55-429e-a52e-dd497bc61664.png)

**Reply by keiji** [reply]
@okumuralab 
ありがとうございます。減少しない点については少し気になりますね……可能であれば動作ログをいただけないでしょうか。

積分表示になった。iOSの仕様ということなら問題はないのですが、1つの可能性としてはCOCOAの機能不全が考えられるので、場合によってはIssueにしておこうと思います。

ある診断キーファイル（zip）があるとき、Androidでは一度接触確認をしたことのある診断キーを与えても反応しませんが、iOSでは確認済みかどうかは判断せず、何度でも接触確認ができます。
それを防ぐために接触確認済みの診断キーのタイムスタンプを保存しておき(LastProcessTekTimestamp)、新しい診断キーだけ接触確認をするようにしています。

この仕組みがうまく働いていないと、接触確認の度に公開されているすべての診断キーの接触確認を行うことになるので、お知らせ頂いたような現象が起きる可能性があります。

動作ログに「No New diagnosis-key found on」という行があるかどうかでは、フィルターが働いているか、いないか確認可能です。
いまぼくの手元で試した限りでは、LastProcessTekTimestampは正常に保存されていて、フィルタリングも行われているのですが…

**Reply by okumuralab** [reply]
動作ログです：

[cocoa_log_6eec3453-39d0-4cd5-b1cc-04e643724015.zip](https://github.com/cocoa-mhlw/cocoa/files/8422819/cocoa_log_6eec3453-39d0-4cd5-b1cc-04e643724015.zip)

**Reply by keiji** [reply]
ありがとうございます。
フィルタリングは問題なく動作していますね。となると本当に積分になっているのかな……

v2.0.0のリリースに当たっては、このあたりの挙動を注視しておきます。万が一LastProcessTekTimestampでのフィルタリングが働かない瞬間があるとすれば、v2.0.0への置き換わりに比例してCDNの負荷が上がるはずです。

また個人的には #928 の対応を入れて、利用者が自分の接触データを自由に確認できるようにするという方向でも調整したいですね。

**Reply by b-wind** [reply]
今の所想定通りに動いてはいるようなので優先度的には低いかもですが、OS側ログの出力内容自体はAppleに問い合わせるなり
テスト環境が用意できるのであればある程度検証可能なのかなとは思います。

**Reply by keiji** [reply]
ひとまずCOCOAの動作不具合を疑って調査するIssueを立てました。 #961

あとは実際に接触状態を作ってMatchCountが積算になるか確認します。

**Comment by b-wind** [comment]
openbeta に限らず、の話になりますが。
ENS v2 への移行により、 iOS の対応バージョンが 13.7 以降になると認識しています。


~~これは急ぎでは無い物の記載しておいた方が良い内容かと思います。~~
流石にβユーザーに該当者は居なさそう。

**Reply by keiji** [reply]
COCOA2のリリース時にバージョンが変わることは明示した方がいいですね。
ありがとうございます。確認しておきます。

**Comment by okkun-zzz** [comment]
Bluetoothをオフにしていると1日1回のキーのダウンロードが行われないようです。
オフの状態ではアマビエさんのスプラッシュスクリーンが表示された後、「確認をして利用する」をタップしてもダウンロードは行われませんでした。
Bluetoothをオンにすれば前回のキーの取得後からその時点での最新のキーまでダウンロードしてきます。
デバイス: Essential Phone PH-1 (Android 10)

v1.4.1 (デバイス: Pixel 3 (Android 12))ではBluetoothがオフの状態でもキーのダウンロードが行われ接触の有無の確認が実行されていました。

**Reply by keiji** [reply]
ご報告ありがとうございます！　これはどういう動きが良いか、ちょっと考えます。

COCOA2は、バックグラウンドの接触確認の最初で接触通知API（EN）が有効であるかをチェックしています。そしてENは、Bluetoothが無効だとENも無効であると判断します。なので、Bluetoothが無効だと、ENそのものが動いていないとしてキーのダウンロードと接触確認を行わないようになります。

対応策を検討します。
まず、Bluetoothが無効の場合も接触確認ができるかを確認します。もし接触確認ができない場合は、現状の動きを維持する（Bluetooth有効化のお願いはシステムが出します）。
もしBluetoothが無効でも接触確認ができる場合は、ENが無効である理由を確認する。Bluetooth無効が原因であれば接触確認は完了するように調整できます。

アプリの開発者としてはBluetoothは常に有効にしておいて欲しいという気持ちがありますが、ご指摘のようにBluetoothを普段無効にしている人もおられるようなので、適切な案内、利用者の利益が最も大きくなる動きにしたいと考えています。。

**Reply by b-wind** [reply]
細かい事ですが、接触キーの ~~ダウンロード~~ 更新は一日最大二回（2/2 頃より）ですね。

ずっとBluetoothが無効の場合は接触の記録自体が行われていないと判断しても良いと思いますが、
一時的にBluetoothを無効にしている（機内モードとかも該当する？）だけの場合、キーのダウンロードと確認は行えた方が良さそうです。

ENS の制限等で難しいようで有れば、その旨通知で出すなども考えられますね。

**Reply by keiji** [reply]
はい。診断キーの更新は、1日に2回行われるようになっていますね。また、これは1日4回に増やしたいという話をしています。

今回、COCOA2からはAndroidでもiOSでも4時間毎にバックグラウンド処理を起動するように変更しています。なので、確認自体は1日6回行うことになります。
これまでAndroidは6時間毎、iOSは4時間毎とずれていました。バックグラウンド処理の間隔を短くすることで、Androidでの接触確認の成功率が向上するのを期待できます。

また、これまではプラットフォーム毎に実行間隔が違っていたので、サーバーのアクセス回数を元にアクティブ数を算出とかやりづらかったのですが、間隔を合わせることで単純に１日のアクセス総数を6分の1するなどしてある程度の精度が見込めるのではないかと期待しています。

**Reply by b-wind** [reply]
診断キーの更新頻度が増えることは良いことだと思います。通知までの時間を抑えることになると思うので。

ネットワーク屋さんとしてはダウンロードの頻度が高くなると、サーバー負荷やユーザーの負担する通信料が少し気になる所ではあります。

**Reply by keiji** [reply]
#934 にIssue作りました！

**Reply by keiji** [reply]
> ダウンロードの頻度が高くなると、サーバー負荷やユーザーの負担する通信料が少し気になる所ではあります。

ありがとうございます。もともとユーザー数の多いiOSが約４時間間隔を目安に設計されていること。日本だと定額通信が一般的なので、ユーザー側の負担が極端に増えることはないだろうと考えていますが、GitHubやストアのレビューを注視していきます。

サーバーの負荷については、スケールできるとは言え直接費用に影響があるので結構シビアですね。
診断キーとlist.jsonのダウンロードの状況を見ながら、場合によっては以前から提案を頂いているetagの導入も視野に入ってくるかと思います。

**Reply by b-wind** [reply]
βの話題からはだいぶズレてしまうのですが。

> 日本だと定額通信が一般的なので、ユーザー側の負担が極端に増えることはないだろうと考えていますが

スマホが対象なので、段階式定額制と捉えていますが、あくまで線形の増加であって激増することは考えにくいですね。
特に契約する通信量を使い切った、所謂「ギガが足りない」人に取ってCOCOAアプリの処理時間は長くなると捉えています。

> 以前から提案を頂いているetagの導入も視野に入ってくるかと思います。

サーバー側の情報更新頻度を上げることも考えると、list.json の圧縮転送を考慮された方が効果が大きいと思います。
（Etag は更新頻度が低いときに有用な手段なので）
https://github.com/cocoa-mhlw/cocoa/issues/199#issuecomment-860258864 に散発的に書いたので別 Issue に纏めた方が良いかな……

**Reply by keiji** [reply]
`list.json`の圧縮いいですね。Issueは分けた方が良さそうですが、クライアント側の実装はぼくの方で試すことができます。
加えて、サーバーとクライアントの間にいるCDNがどういう挙動をするか分かってないので、サーバーチームにも聞いてみます。

**Reply by b-wind** [reply]
圧縮転送の提案に付いては #938 に簡潔ですが纏めておきました。

**Comment by b-wind** [comment]
COCOA v2.0.0-openbeta から「設定」項目にオープンソースライセンスという項目が増えていますね。（クローズドβ版は見てないです）
「COCOA はオープンソースソフトウェアです。Mozilla Public License 2.0 の条項下で提供されています」
との記述はありますが、MPL v2 ライセンス条項の表示やリンク等が無いのでどのような条項があるのか分かりづらい状況になっています。
即ダメという訳では無いかもですが、あまり嬉しくは無い表記かなと。


また、「ライセンス表記」というボタンを押すとCOCOAとは一見関係無いソフトウェアの表示が並びます。
こちらもボタン名を適切な名前にする必要性を感じます。

**Reply by keiji** [reply]
そうですね。ご指摘の「ライセンス表記」は本来的には「オープンソースライセンス」と書いた方がいいと思います（それでも一般利用者には何のことか分からない人が居ると思いますが）。

一方、「オープンソースライセンス」にすると横に長くなるので収まりが悪く、これまでの表記を維持しています。アプリ開発者からすると、上部バーメニュー内に「オープンソースライセンス」の名称で置きたいのですが、アクセシビリティ的には上部バーのメニューは使わない方針なので、どのように取り扱うかは結構悩ましいところです。
最近、使い方がメニューからドロワー内に移動したり、GitHubのリンクを開く操作が「https://...」から始まるリンクテキストでなく、ボタンとして独立しているのも同様の理由です。

"3.1 Distribution of Source Form"はソース形式で配布する場合ですね。COCOAのアプリ（実行形式）としての配布は "3.2 Distribution of Executable Form" が適用されて、ソースコードの入手方法が案内されているという条件をクリアすることを念頭に今の形にしています。
こちらもわかりにくいよってのは理解できるので、もっと良い形を模索していきたいと考えているので、アイデアをもらえるとありがたいです。

**Reply by b-wind** [reply]
個人的にはCOCOA自体のライセンスの説明と利用しているライブラリ等のライセンスの表示が混ざっているのが分かりにくいかなと。
たたき台としては以下の感じではどうでしょうか。


## COCOAライセンス

COCOA はオープンソースソフトウェアです。
Mozilla Public License 2.0 の条項下で提供されています
COOCA のソースコードは以下のURLから入手出来ます。

https://github.com/cocoa-mhlw/cocoa/

Mozilla Public License 2.0 の条項は以下のURLから確認出来ます。

https://www.mozilla.org/en-US/MPL/2.0/

- 「ソースコードを入手」ボタン
- 「ライセンスを確認」ボタン

## オープンソースライセンス

- 「ライセンスを確認」ボタン

**Reply by b-wind** [reply]
MPL 原文はアプリ内で持って、直接表示する方が良いですね。リンク先でもほぼ変わることは無いとは思いますが。

**Reply by keiji** [reply]
ありがとうございます。COCOA自体のライセンスとオープンソースライセンスを分けるというのは良いアイデアだと思います。
#941 でIssue化しました！
